frontends/pytorch/examples/torchfx2iree/builder.py

```diff
@@ -0,0 +1,101 @@
+# -*- Python -*-
+# This file is licensed under a pytorch-style license
+# See frontends/pytorch/LICENSE for license information.
+#
+# pylint: disable=no-member, no-name-in-module, invalid-name
+
+from typing import Mapping, Optional
+
+from npcomp import ir
+from npcomp.dialects import builtin, std
+import npcomp.dialects.torch as torch_d
+from torch_mlir import ModuleBuilder
+import torch.fx
+
+Environment = Mapping[torch.fx.Node, ir.BlockArgument]
+
+
+class Builder:
+    def __init__(self) -> None:
+        self.module_builder = ModuleBuilder()
+        self.module = self.module_builder.module
+        self.module_ip = ir.InsertionPoint(self.module.body)
+        self.context = self.module.context
+        self.env: Environment = {}
+        self.graph_module: Optional[torch.fx.GraphModule] = None
+        # TODO: find a way to get a real location
+        self.loc = ir.Location.unknown(self.context)
+
+    def import_module(self, m: torch.fx.GraphModule) -> ir.Module:
+        self.graph_module = m
+        tensor_type = self.get_torch_tensor_type()
+        # TODO: currently I am assuming that all inputs are tensors
+        f_type = ([tensor_type] * self.get_num_of_args(), [tensor_type])
+        with self.context:
+            f = builtin.FuncOp("forward", f_type,
+                               loc=self.loc,
+                               ip=self.module_ip)
+
+        f.add_entry_block()
+        f_ip = ir.InsertionPoint(f.entry_block)
+        self.initialize_environment(f.entry_block.arguments)
+
+        for node in m.graph.nodes:
+            if node.op == 'call_function':
+                result = self.insert_function_call(node, ip=f_ip)
+                self.env[node] = result
+            if node.op == 'output':
+                std.ReturnOp([self.env[node_arg] for node_arg in node.args],
+                             loc=self.loc, ip=f_ip)
+        return self.module
+
+    def insert_function_call(self, f_node: torch.fx.Node, *,
+                             ip: Optional[ir.InsertionPoint] = None
+                             ) -> ir.OpResult:
+        assert f_node.op == 'call_function'
+        args = [self.env[arg_node] for arg_node in f_node.args]
+
+        if f_node.target == torch.sigmoid:
+            assert len(args) == 1, "`sigmoid` only takes one argument"
+            result = torch_d.AtenSigmoidOp(self.get_torch_tensor_type(),
+                                           args[0],
+                                           loc=self.loc,
+                                           ip=ip).result
+        elif f_node.target == torch.sin:
```

Comment:
If you end up keeping this big if/else chain, probably better to hoist it to a lookup table that resolves to a callback of some kind. Various ways to structure that. For extensibility, I somewhat prefer the decorator approach:

```
TARGET_BUILDERS = dict()

def target_builder(f=None, *, target):
  if not f: return functools.partial(target=target)
  def wrapped(...):
    TARGET_BUILDERS[target] = f
    return f

@target_builder(target=aten.sin)
def aten_sin_builder(...)
  ...
```

May need fiddling but something like that.